packages/server/src/threads/automation.ts

```diff
@@ -330,286 +310,224 @@ class Orchestrator {
     loopStep: LoopStep,
-    steps: AutomationStep[],
-    stepIdx: number,
-    pathIdx?: number
-  ): Promise<number> {
-    await processObject(loopStep.inputs, this.mergeContexts(this.context))
-    const iterations = getLoopIterations(loopStep)
-    let stepToLoopIndex = stepIdx + 1
-    let pathStepIdx = (pathIdx || stepIdx) + 1
-
-    let iterationCount = 0
-    let shouldCleanup = true
-    let reachedMaxIterations = false
-
-    for (let loopStepIndex = 0; loopStepIndex < iterations; loopStepIndex++) {
-      try {
-        loopStep.inputs.binding = automationUtils.typecastForLooping(
-          loopStep.inputs
-        )
-      } catch (err) {
-        this.updateContextAndOutput(
-          pathStepIdx + 1,
-          steps[stepToLoopIndex],
-          {},
-          {
-            status: AutomationErrors.INCORRECT_TYPE,
-            success: false,
-          }
-        )
-        shouldCleanup = false
-        break
-      }
-      const maxIterations = automationUtils.ensureMaxIterationsAsNumber(
-        loopStep.inputs.iterations
-      )
-
-      if (
-        loopStepIndex === env.AUTOMATION_MAX_ITERATIONS ||
-        (loopStep.inputs.iterations && loopStepIndex === maxIterations)
-      ) {
-        reachedMaxIterations = true
-        shouldCleanup = true
-        break
-      }
-
-      let isFailure = false
-      const currentItem = this.getCurrentLoopItem(loopStep, loopStepIndex)
-      if (currentItem && typeof currentItem === "object") {
-        isFailure = Object.keys(currentItem).some(value => {
-          return currentItem[value] === loopStep?.inputs.failure
-        })
-      } else {
-        isFailure = currentItem && currentItem === loopStep.inputs.failure
-      }
+    stepToLoop: AutomationStep
+  ): Promise<AutomationStepResult> {
+    await processObject(loopStep.inputs, prepareContext(ctx))
+
+    const result = {
+      id: loopStep.id,
+      stepId: loopStep.stepId,
+      inputs: loopStep.inputs,
+    }
 
-      if (isFailure) {
-        this.updateContextAndOutput(
-          pathStepIdx + 1,
-          steps[stepToLoopIndex],
-          {
-            items: this.loopStepOutputs,
-            iterations: loopStepIndex,
-          },
-          {
-            status: AutomationErrors.FAILURE_CONDITION,
-            success: false,
-          }
-        )
-        shouldCleanup = false
-        break
-      }
+    const loopMaxIterations =
+      typeof loopStep.inputs.iterations === "string"
+        ? parseInt(loopStep.inputs.iterations)
+        : loopStep.inputs.iterations
+    const maxIterations = Math.min(
+      loopMaxIterations || env.AUTOMATION_MAX_ITERATIONS,
+      env.AUTOMATION_MAX_ITERATIONS
+    )
 
-      this.context.steps[pathStepIdx] = {
-        currentItem: this.getCurrentLoopItem(loopStep, loopStepIndex),
+    const items: Record<string, any>[] = []
+    let iterations = 0
+    let iterable: any[] = []
+    try {
+      iterable = getLoopIterable(loopStep)
```

Comment:
What is this? The response type appears to be `any` and the loop below makes it look like a number, but the name `iterable` suggests its some sort of array/iterable object

Reply:
The response type is `any[]`.